### ISSUE TITLE
NPC

### ISSUE THREAD

**Issue by Phoenix616** [Minebench/Tresor#23]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%240%20Funded-%2300A156.svg" />](https://oss.issuehunt.io/r/Minebench/Tresor/issues/23)

Design a service provider for working with NPCs.

Questionable how configurable this should be. But at least creating simple ones and querying for existence might be a good start.

Plugins that will be supported right out of the gate:

- Citizens
- find more?

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Submitted pull Requests
- [#26 NPC Draft](https://oss.issuehunt.io/r/Minebench/Tresor/pull/26)
---

### Tips

- Checkout the [Issuehunt explorer](https://oss.issuehunt.io/r/Minebench/Tresor/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://oss.issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by IllusionTheDev** [comment]
Different NPC systems have different ways of approaching NPC IDs. I'm questioning if I should make a generic NPCIdentifier interface to account for that (Citizens uses numerical IDs, other plugins may use UUIDs or internal names)

**Comment by Phoenix616** [comment]
Yeah that's indeed an interesting question. This might actually be the case for more services than just this now that I think about it (e.g. someone might not use a string for the holograms either) so maybe having a general Identifier interface in the Tresor API that can be used by any service which might handle IDs could be best (one could even argue that a player using a UUID is kinda arbitrary and should use such an interface...)

If you want to take a stab at designing something like that feel free but I will too have to think about the best approach for that as it needs to be user-friendly enough to be easily usable. (E.g. you wouldn't want to have to always wrap your player-UUID to use an Economy service but this might not be that big of an issue in the case of NPCs or even holograms where you already have the Identifier from creating it. But again, some people might want to store them in a config/database, so any Identifier implementation would also need to be able to properly be converted into a storeable format...)

So yeah, thanks for bringing this up. If you have a good idea for addressing these points then feel free but it's definitely not part of this issue and if you want to do NPCs too then feel free to just implement it in a way that you think should work with most (e.g. [Citizens can get NPCs via UUID](https://jd.citizensnpcs.co/net/citizensnpcs/api/npc/NPCRegistry.html#getByUniqueId(java.util.UUID)) too, maybe just using that for now would be the best approach)

**Comment by IllusionTheDev** [comment]
Another issue I can think of is persistence. Certain APIs create persistent / non-persistent NPCs/Holograms etc. How should we tackle this? Citizens has its NPCRegistry system that allows us to choose, I'm unsure about other plugins like ZNPCs

**Comment by Phoenix616** [comment]
Sounds like something that could be put behind a Feature enum (like one for persistent, one for temporary) and an option on NPC creation to store it or not. (And a create method without a parameter to be used with all plugins which should probably default to storing for those that can do both as I would assume that's what one usually expect it to do)

**Comment by Asobu01** [comment]
Happy to own this bounty. I’ll keep updates transparent, scope focused, and deliver a clean PR with validation notes within 24h. I optimize for first-pass acceptance by keeping changes minimal, auditable, and aligned with existing code style. I am currently tracking 26 active bounty thread(s) with timely updates. (@Asobu01)

**Comment by Asobu01** [comment]
Quick follow-up on `Minebench/Tresor#23`. If this issue is still open, I can start right away and post a short execution checklist first. (@Asobu01)

**Comment by ysliaoniubi** [comment]
Withdrawing our attempt: the maintainers closed AI-generated PRs per the contributor agreement. We'll refrain from further submissions on this issue.

**Comment by Roc755** [comment]
👋 I've started implementing the NPC service from scratch (方案A). Currently working on Citizens provider integration. Will submit PR once ready. ETA: within a few hours.
